x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,434 @@
+     *                    Id Token as allowed in the specification. In such a case we can't validate the access token
+     *                    but this is considered safe as it was received in a back channel communication that was protected
+     *                    by TLS.
+     */
+    private void validateAccessToken(AccessToken accessToken, JWT idToken, boolean optional) {
+        try {
+            // only "bearer" is defined in the specification but check just in case
+            if (accessToken.getType().toString().equals("Bearer") == false) {
+                logger.debug("Invalid access token type [{}], while [Bearer] was expected", accessToken.getType());
+                throw new ElasticsearchSecurityException("Received a response with an invalid access token type");
+            }
+            String atHashValue = idToken.getJWTClaimsSet().getStringClaim("at_hash");
+            if (null == atHashValue && optional == false) {
+                logger.debug("Failed to verify access token. at_hash claim is missing from the ID Token");
+                throw new ElasticsearchSecurityException("Failed to verify access token");
+            }
+            AccessTokenHash atHash = new AccessTokenHash(atHashValue);
+            JWSAlgorithm jwsAlgorithm = JWSAlgorithm.parse(idToken.getHeader().getAlgorithm().getName());
+            AccessTokenValidator.validate(accessToken, jwsAlgorithm, atHash);
+        } catch (Exception e) {
+            logger.debug("Failed to verify access token.", e);
+            throw new ElasticsearchSecurityException("Failed to verify access token.");
+        }
+    }
+
+    /**
+     * Parses and validates an OpenID Connect Id Token to a set of claims
+     *
+     * @param idToken       The Id Token to parse and validate
+     * @param expectedNonce The nonce that was generated in the beginning of this authentication attempt and was stored at the user's
+     *                      session with the facilitator
+     * @return a {@link JWTClaimsSet} with the OP claims that were contained in the Id Token
+     */
+    @SuppressForbidden(reason = "uses toFile")
+    private JWTClaimsSet validateAndParseIdToken(JWT idToken, Nonce expectedNonce) {
+        Secret clientSecret = null;
+        try {
+            final IDTokenValidator validator;
+            final JWSAlgorithm requestedAlgorithm = rpConfig.getSignatureVerificationAlgorithm();
+            if (JWSAlgorithm.Family.HMAC_SHA.contains(requestedAlgorithm)) {
+                clientSecret = new Secret(rpConfig.getClientSecret().toString());
+                validator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(), requestedAlgorithm, clientSecret);
+            } else {
+                String jwkSetPath = opConfig.getJwkSetPath();
+                if (jwkSetPath.startsWith("https://")) {
+                    validator = new IDTokenValidator(opConfig.getIssuer(), rpConfig.getClientId(), requestedAlgorithm,
```

Comment:
If we want to use this, we need to fork to a different threadpool as they expect a synchronous resolution or go async to retrieve the content ourselves and then use `JWKSet.parse` on the returned body content

Reply:
I think we would have to do some caching to not reach out for `JWKSet` for each new authn. For that matter I think it's better to let the `IDTokenValidator` handle its business on a separate thread and not build the `IDTokenValidator` everytime (make it an instance variable).

Reply:
Good point regarding making IDTokenValidator an instance variable. It is thread safe. I do wonder if we should have some sort of built in update mechanism for obtaining the jwkset rather than having the token validator reach out for every auth.

Reply:
I moved `IDTokenValidator` an instance variable and since it's thread safe I think we'll be fine. WDYT ?
 It won't reach out for the jwkset on each authn, as there is a simple built in caching mechanism. It will only reach out again, in case it receives a JWT that carries a `kid` in its Header that the `IDTokenValidator`  has not seen before ( the use case is OP key rotation )

Reply:
I also added a `FileWatcher` to make sure that we pick up changes to the JWKSet if it's provided in a local file